manipulation.go

```diff
@@ -572,3 +652,24 @@ func (s *Selection) manipulateNodes(ns []*html.Node, reverse bool,
 
 	return s
 }
+
+func (s *Selection) eachNodeHtml(htmlStr string, parent bool, fn func(n *html.Node, nodes []*html.Node)) *Selection {
+	nodesMap := make(map[html.NodeType][]*html.Node)
+	var context *html.Node
+	for _, n := range s.Nodes {
+		if parent {
+			context = n.Parent
+		} else {
+			context = n
+		}
+		if context != nil {
+			nodes, found := nodesMap[context.Type]
```

Comment:
The cache isn't quite right - it caches the parsed nodes based on the context node's type (TextNode, ElementNode, etc.) but the parser is more fine-grained than that - e.g. it won't parse a `"<tr><td>...</td></tr>"` the same way inside a `<table>` vs inside say a `<div>`. I think a better way would be to store it based on the equivalent of calling `goquery.NodeName` (except that this `NodeName` function should be refactored to call an unexported function that accepts an `*html.Node` so that it is faster when we already have a node).

Here's an example code that illustrates the issue (should probably be turned into a test case), in `d3`, the call to `AppendHtml` will add the `<tr>...</tr>` nodes both inside the `<table>` (which is fine) and the `<a>` (which isn't):

```
var insideA = `<html>
<body>
	<a class="x"></a>
</body>
</html>`

var insideTable = `<html>
<body>
	<table class="x"></table>
</body>
</html>`

var insideBoth = `<html>
<body>
	<table class="x"></table>
	<a class="x"></a>
</body>
</html>`

func main() {
	d1, err := goquery.NewDocumentFromReader(strings.NewReader(insideA))
	if err != nil {
		log.Fatal(err)
	}
	d2, err := goquery.NewDocumentFromReader(strings.NewReader(insideTable))
	if err != nil {
		log.Fatal(err)
	}
	d3, err := goquery.NewDocumentFromReader(strings.NewReader(insideBoth))
	if err != nil {
		log.Fatal(err)
	}

	s1 := d1.Find(".x").AppendHtml("<tr><td>Hello</td></tr>")
	s2 := d2.Find(".x").AppendHtml("<tr><td>Hello</td></tr>")
	s3 := d3.Find(".x").AppendHtml("<tr><td>Hello</td></tr>")

	o1, _ := goquery.OuterHtml(s1)
	o2, _ := goquery.OuterHtml(s2)

	fmt.Println(o1, "\n\n\n")
	fmt.Println(o2, "\n\n\n")

	s3.Each(func(i int, s *goquery.Selection) {
		o, _ := goquery.OuterHtml(s)
		fmt.Println(o, "\n\n\n")
	})
}
```